### ISSUE TITLE
Design interaction of data exporting

### ISSUE THREAD

**Issue by noahmanger** [18F/openFEC-web-app#946]
**Stories:** So that users can export data from the site, make it possible to:
- [Story 1](https://18f.storiesonboard.com/m/fec/!card/5wzt02F1sSSdiTnUBqI-VQ): ...instantly download a small set of filtered data.
- [Story 2](https://18f.storiesonboard.com/m/fec/!card/VA97jHrDIs-ibznUBqI-VQ): ...wait on the site for large datasets to be ready.
- [Story 3](https://18f.storiesonboard.com/m/fec/!card/5wdNOSHSZnmJ2znUBqI-VQ): ...find relevant bulk downloads if that's most helpful. 

Task: Design the interactions for these scenario. Features to include / questions to answer:
- Provide tips to reduce the size of the query in order to get a faster download
- Alert state in case the request is too large and can't be exported at all
- Alert state if request is empty
- Does the waiting option provide a link to a different page? Is it a persistent element on the site?

**Comment by noahmanger** [comment]
Another question to answer, coming from @onezerojeremy 's notes on usability testing: does the export include _all_ possible columns? Is there a step where users select the columns they want to export? Does it just export what's currently visible on the table?

@jmcarp and @LindsayYoung are there any backend considerations to account for here? My guess is the easiest solution for the first version of this is probably to just have it export _all_ columns, and maybe down the road add the option to customize?

**Comment by LindsayYoung** [comment]
I think it would be good to have the ability to export all the columns, including fields we don't show in the viewer. That way, if you are an advanced data user without technical skills, you can still have access to the data.

I like the idea of starting with uniform fields for downloads and then making them custom. Making smaller requests, could speed things up.

**Comment by noahmanger** [comment]
Another interaction question: do we want to support downloading in multiple formats? Or just CSV?

**Comment by noahmanger** [comment]
So here's a flow I'm thinking (with very ugly screens; just meant for general page components, not actual style):

From a browse page, clicking the download button takes you directly to a new page in a new tab. If the request is small enough, the download button will be ready for you. Otherwise it will display a processing message:

![download waiting](https://cloud.githubusercontent.com/assets/1696495/11289245/b4a61d3c-8edf-11e5-827c-8e2894222630.png)

On that page you can see the filters applied as well as the total number of results. You can also cancel the download if you want. This page lives on a unique URL that is accessible for some amount of time. My thought is that all download requests, both big and small, should send the user to this page. That way it an be shared with others or to the user their self to download later.

This page also contains instructions for how to get faster results and how to find bulk downloads.

Once it's ready, the button becomes active:

![download ready](https://cloud.githubusercontent.com/assets/1696495/11289276/efdaac24-8edf-11e5-8b0c-3fdd2b4b7b2f.png)

Not necessary, but it could be cool to have a persistent bar while the system prepares the download that persists across screens on the site and notifies the user when it's ready:

![waiting](https://cloud.githubusercontent.com/assets/1696495/11289289/197a7898-8ee0-11e5-956d-1dcb6004ed23.png)
![ready](https://cloud.githubusercontent.com/assets/1696495/11289290/197ab98e-8ee0-11e5-8fd2-3274c6741157.png)

I'm looking for feedback on both the experience of the flow (ping @jenniferthibault and @onezerojeremy ) as well as the technical implications of this. Some technical questions I've got right now:

- At which point do we know if a query is too big to download? Would it be possible to not even have the download button active on the main browse page if the results are above a certain amount? Or do we need to request the download and then have an warning message on the download page?
- Is it possible to have this kind of persistent bar?
- Does it make sense to show the filters applied on this page? I thought it would be valuable so that if you return to this page you see what the download is going to contain.
- Should the download page have a timestamp?

Kind of a braindump, but wanted to share this with folks.

**Comment by jenniferthibault** [comment]
Haven't dug into mockups yet, but on the file type info, I just reviewed all submitted feedback issues that requested this: 

Out of the 9 issues
- 7 of them specifically asked for .csv/Excel
- 1 asked for "pdf/html/FEC files" 
- 1 asked for "data export" no format specified

**Comment by noahmanger** [comment]
Good call. I believe that the "PDF/HTML/FEC files" is specifically about the filings, not exporting data. I think we can just count on a CSV.

**Comment by onezerojeremy** [comment]
I like this direction! I especially like the idea of a bar showing that the download is processing. Is it possible to just do that? Probably not MVP. 

I'm also curious about what happens with the bar if you have multiple downloads processing (use case: comparing two candidates spending/fundraising)

**Comment by onezerojeremy** [comment]
Also, from testing: 

One test subject was taken aback by the thought of having to wait ("torrents are immediate"), and even more disturbed by the idea of having to enter an email address. This lends further credence to the current interaction pattern that does not require an email address. We should think about how to explain what is happening in a bit more detail; people expect instantaneous downloads, which we in fact provide, but each download requires cobbling together database entries first. Consider more specificity and concision in description of system status.   
One individual found the term CSV to be too technical, they had a vague notion that it was something that could be loaded into a spreadsheet, but weren't totally clear. Consider language change from "Export CSV". Icons were also suggested.

**Comment by noahmanger** [comment]
Great stuff! I like the idea of going from "Button click" --> "Banner appears" and then you can click to that page if you want or just keep doing what you're doing. 

As for multiple downloads, that's a really good point. I'm guessing each one needs a different page @jmcarp / @LindsayYoung ? If so, I think we could just stack the download links, with the ability to hide the whole bar:

So you initiate a download here:

![browse page](https://cloud.githubusercontent.com/assets/1696495/11344638/648a57d8-91c7-11e5-8ab8-60d40054aec7.png)

And then the bar slides down on that page and you can keep doing what you're doing:

![browse - pending](https://cloud.githubusercontent.com/assets/1696495/11344644/72d4a79e-91c7-11e5-82ff-d5c25df14b7d.png)

**Comment by noahmanger** [comment]
Summary of our discussion today, we're going to break this interaction up into a couple versions. 

## V1: Notification bar only
- Both instant and queued downloads will cause the notification bar to appear
- Once the download is ready, a link to the csv will appear in that bar 
- The bar will contain information on A) what's happening while the download processes and B) a suggestion to add more filters to get quicker downloads in the event of a queued download
- We will save the download URLs in user's localstorage so that when they return to the site on the same device, we will show the status of their download

## V2: Notification bar + waiting page
- The next iteration will add a link to a unique "waiting page" in the notification bar for both instant and queued downloads.
- The waiting page will contain the information from the mocks above, as well as further description of the columns contained in the download

I'm adding this as a separate story card.

## Remaining tasks for v1
- [ ] @emileighoutlaw draft content
- [ ] @jenniferthibault refine design
- [ ] Josh will get things going on the front end
- [ ] I'll add styles

Shooting to have this deployed for testing this week.

**Comment by emileighoutlaw** [comment]
>Download in progress: XXXXXXX
>Taking too long? *Narrow your search*[link] to speed this up.

Questions:
- @LindsayYoung or @jmcarp: Can you help me understand what is happening technically and in rough terms, and I'll finesse?
- Can we (should we?) add an "estimated time left"?

**Comment by jenniferthibault** [comment]
This is going to be a few steps of explanation—bear with me :) 

### The export button: 
@emileighoutlaw in these mockups I was making a clear distinction between "export" and "download" actions. For example, the button that starts the process of gathering data is "Export" and the button that allows you to download the file when it's ready is "Download". Does this make sense or would you rather change them to something else?

@noahmanger here are the changes to the layout of the browse table header area that will clean it up and make room for the button:
![screen shot 2015-12-01 at 12 14 16 am](https://cloud.githubusercontent.com/assets/11636908/11492858/8099ade4-97c0-11e5-8cc7-999650db6cac.png)

![screen shot 2015-12-01 at 12 14 01 am](https://cloud.githubusercontent.com/assets/11636908/11492859/8abcf128-97c0-11e5-9ecc-94f8d079709d.png)
 - Can we move the `Hide results with missing values when sorting` checkbox into the filter panel? Even if it's selected and applied by default, it seems to make more sense there than in this header space, which seems to be more for navigation.

### The Notification bar  
2 options, with different content. Not sure how much is possible, so the constraints will likely choose the design to go with. I've moved all the bars to the bottom, instead of the top of the screens, so that they will interfere with regular navigating less. The trade-off is that they block the feedback tool, but I think that's ok for this short amount of time. (?)

#### Option 1  

It would be neat if there could be a dynamic change on the bar when it's ready for download, and 3 is highlighted.  

**WAITING**
![waiting](https://cloud.githubusercontent.com/assets/11636908/11492902/15c3c850-97c1-11e5-85af-28e8746ee19b.png)

**READY**
![ready](https://cloud.githubusercontent.com/assets/11636908/11492992/fdeeb13a-97c1-11e5-9b6c-1e179b9f1524.png)

#### Option 2  

**WAITING** 
![waiting copy](https://cloud.githubusercontent.com/assets/11636908/11492987/ed1e1cc4-97c1-11e5-9b87-67169fdbd5c7.png)

**READY**
![ready copy](https://cloud.githubusercontent.com/assets/11636908/11492988/f52650f8-97c1-11e5-8eb0-130292f7d8ed.png)


One more question: Do we need to put an x in the corner of this notification bar, or other way to close it? This could allow users to cancel an export in progress, but I'm not sure if that's desired functionality.

**Comment by noahmanger** [comment]
Nice nice nice. Thoughts:

- Love the distinction between "export" and "download". That's smart.
- I think I prefer the first option for the status bar
- It feels a little weird to me at the bottom of the screen. What if we put it at the top but it actually pushed everything down, rather than overlapping?
- I think a "cancel" link would probably be good, but I think it should be distinct from an "x" or "Hide" button. Not sure if we want one of those as well.
- I like the redesign of the navigation controls. I'm not sure about the "hide null while sorting" button. It's not actually a filter, it's almost more of a navigation control. Maybe it could be in the controls at the bottom of the table?
- Once you initiate an export, what state happens to the original export button? Do we want a "loading" state or something? Or just disabled until you change filters?
- You're introducing a new button style with the beige text on dark blue background. Previously we were using aqua against blue. Do you want to do this? What are the rules for when each is used?

**Comment by emileighoutlaw** [comment]
These are some interesting questions:

### Export and download
As a reader, I think `export` and `download`  are good, clear terms. 

Before we decide, though, I'd be curious to hear if Lindsay or Josh think there's a more technically appropriate word for either of these things. I'm still not 100% sure what's happening when the user clicks "export" that's different than when they click "download."

### Option 1
I like that users get a visualization of the steps. But step 1 is completed *before* the user clicks "export," so I'm not sure it makes sense to include it. Overall, it feels a little text heavy to me.

### Option 2
I prefer this one. I'm drawn to the simplicity— I think it's just as clear as the first one as to what's happening, and I'm in favor of the light copy.

#### Minor content thought on Option 2
I usually dislike elipses, but I think `Export in progress...` would imply a nice motion here. And then, no punctuation after `Download ready` for a more friendly tone.

Also, what do you think about `Add more filters for faster exports` as the word ordering— to more quickly get to the meat of what users should do.


### Cancel download
I do like adding a cancel option, but I don't know if folks will intuitively know X-ing out of the box will cancel the download (as opposed to just hiding the box). I don't want to overstep into design specifics, but in my own web interactions, I like seeing things like: 

![images-1](https://cloud.githubusercontent.com/assets/10552702/11512976/d5deebba-983e-11e5-806e-5c56965f11ae.jpg)
![images](https://cloud.githubusercontent.com/assets/10552702/11512978/d837d58e-983e-11e5-953e-f151a9a6328b.jpg)

**Comment by jmcarp** [comment]
FWIW, "cancel" is a little complicated. If two users queue the same data download at the same time, we only build the file once. That means multiple users can share the same queued download, which means adding a naive cancel button would allow one user to accidentally cancel downloads requested by somebody else.

**Comment by onezerojeremy** [comment]
i like this direction a lot! 
I've no strong emotions between 1 and 2 ( i like them both). I also don't have strong emotions about top or bottom (i don't know if there are patterns or expectations here) but don't mind bottom. 

one question: what happens if you have multiple downloads at once? 

on cancel:
I think we should interpret "Cancel" here to equivalent of the user saying "I no longer desire this file and would like for this box to go away" (likely such that they can do a different search unimpeded). If multiple people have requested the same file, it certainly shouldn't keep the other user from getting it. That said, if it is a lone request (and perhaps a db-heavy one) it does seem like there's computational value in letting them tell us we can forgo their request). I suspect there's also a computational tax (not to mention code to write) in figuring out which process to cancel, not sure which is greater.

**Comment by LindsayYoung** [comment]
I would guess that actually canceling would be more trouble than it is worth. 

How about "make your request smaller" or "request fewer records" instead of "add more filters" I think smaller=faster will make sense to people.

**Comment by jenniferthibault** [comment]
Alrighty! Here's the latest after chatting desired behavior for draft 1 of this thing with Noah and Josh:

![waiting copy 2](https://cloud.githubusercontent.com/assets/11636908/11541470/68b974e8-9900-11e5-8ef9-81e7c30d5154.png)

- @emileighoutlaw I used your text combined with @LindsayYoung 's about faster downloads. I think it's helpful to include the instruction that you need to use the filters to request fewer records, which in turn builds a faster export process. There's likely some wordsmithing to do here, as well as the `Building your exports...` title, which could be anything from `Building your .CSV` to `Preparing your download` and many other ways to say, "We're making the thing you're waiting for"

- I talked this through with Noah, but for transparency and for others paying attention to this stream, I'm inclined to test the first iteration of this placed on the bottom of the screen to simulate the pattern of where browsers place file download alerts. 

- The X icon to the right of the `Download` button would be to dismiss the download from your queue (not "Cancel" it, per se). This is in line with @onezerojeremy 's logic, which followed what I was thinking as well. 

- For now, not including a "Hide" button for this panel. It will be there until users clear their download queue by using the dismiss buttons.

- Once you initiate an export, I'd like to try the first round just disabling the `Export` button until the user changes filters.

- It no longer hides the feedback button! 

- Download file names will be structured like so `category_mm-dd-yyyy_hh:mm` giving us something like `candidates_12-02-2016_11:49`

- @jmcarp is going to work up a zipped download manifest that will be included with each download, so that users can tell what filters were applied to their .CSV files

Next steps before testing:
- [ ] @emileighoutlaw decides on copy & wordsmithing
- [ ] @noahmanger builds out front end

**Comment by noahmanger** [comment]
This has been designed and implemented. Closing!
